src/backend/distributed/commands/type.c

```diff
@@ -0,0 +1,734 @@
+{
+	const char *alterTypeStmtSql = NULL;
+	TypeName *typeName = NULL;
+	Oid typeOid = InvalidOid;
+	ObjectAddress typeAddress = { 0 };
+
+	Assert(stmt->relkind == OBJECT_TYPE);
+
+	if (IsTransactionBlock())
+	{
+		return NIL;
+	}
+
+	if (ExtensionStmtInProgess())
+	{
+		/*
+		 * extensions should be created separately on the workers, types cascading from an
+		 * extension should therefor not be propagated here.
+		 */
+		return NIL;
+	}
+
+	/* check if type is distributed before we run the coordinator check */
+	typeName = makeTypeNameFromRangeVar(stmt->relation);
+	typeOid = LookupTypeNameOid(NULL, typeName, false);
+	ObjectAddressSet(typeAddress, TypeRelationId, typeOid);
+	if (!isObjectDistributed(&typeAddress))
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	/* reconstruct alter statement in a portable fashion */
+	alterTypeStmtSql = deparse_alter_type_stmt(stmt);
+	ereport(DEBUG3, (errmsg("deparsed alter type statement"),
+					 errdetail("sql: %s", alterTypeStmtSql)));
+
+	/* TODO, mx expects the extension owner to be used here, this requires an alter owner statement as well */
+	EnsureSequentialModeForTypeDDL();
+	SendCommandToWorkersAsUser(ALL_WORKERS, DISABLE_DDL_PROPAGATION, NULL);
+	SendCommandToWorkersAsUser(ALL_WORKERS, alterTypeStmtSql, NULL);
+
+	return NULL;
+}
+
+
+List *
+PlanCreateEnumStmt(CreateEnumStmt *stmt, const char *queryString)
+{
+	const char *createEnumStmtSql = NULL;
+	RangeVar *var = NULL;
+	ObjectAddress typeAddress = { 0 };
+	Oid typeOid = InvalidOid;
+	TypeName *typeName = NULL;
+
+	if (ExtensionStmtInProgess())
+	{
+		/*
+		 * extensions should be created separately on the workers, types cascading from an
+		 * extension should therefor not be propagated here.
+		 */
+		return NIL;
+	}
+
+	/*
+	 * managing types can only be done on the coordinator if ddl propagation is on. when
+	 * it is off we will never get here
+	 */
+	EnsureCoordinator();
+
+	/* enforce fully qualified typeName for correct deparsing and pg_dist_object */
+	var = makeRangeVarFromNameList(stmt->typeName);
+	if (var->schemaname == NULL)
+	{
+		makeRangeVarQualified(var);
+		stmt->typeName = list_make2(makeString(var->schemaname),
+									makeString(var->relname));
+	}
+
+	/* lookup type address of just created type */
+	typeName = makeTypeNameFromNameList(stmt->typeName);
+	typeOid = LookupTypeNameOid(NULL, typeName, false);
+	ObjectAddressSet(typeAddress, TypeRelationId, typeOid);
+
+	EnsureDependenciesExistsOnAllNodes(&typeAddress);
+
+	/* reconstruct creation statement in a portable fashion */
+	createEnumStmtSql = deparse_create_enum_stmt(stmt);
+	ereport(DEBUG3, (errmsg("deparsed enum type statement"),
+					 errdetail("sql: %s", createEnumStmtSql)));
+	createEnumStmtSql = wrap_in_sql(CREATE_OR_REPLACE_COMMAND, createEnumStmtSql);
+
+	/* to prevent recursion with mx we disable ddl propagation */
+	/* TODO, mx expects the extension owner to be used here, this requires an alter owner statement as well */
```

Comment:
`SendCommandToWorkers` uses `CitusExtensionOwnerName()`, we should maybe go for that

Reply:
That could work, but I really like the idea of sending it as the current user in most situations. It makes ACL's applied and we don't have to add an extra command that changes the owner.

It could be that since we don't propagate when in xact it already works with MX as expected now. I will have a look.

Reply:
We should probably hurry to propagate schema grants to avoid situations where we cannot actually create the type on the worker.